WalletWasabi/WabiSabi/Client/GreedyDecomposer.cs

```diff
@@ -0,0 +1,31 @@
+using System.Linq;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class GreedyDecomposer
+	{
+		public GreedyDecomposer(IEnumerable<long> denominations)
+		{
+			_denominations = denominations.Any()
+				? denominations.OrderByDescending(x => x).ToImmutableList()
+				: throw new ArgumentException($"Argument {nameof(denominations)} has no elements");
+		}
+
+		private ImmutableList<long> _denominations;
+
+		public IEnumerable<long> Decompose(long amount) =>
```

Comment:
What is the pattern of usage of this when I have more coins to decompose? Create another GreedyDecomposer and pass the previous result into the ctor?

Reply:
The idea is to decompose the total amount and not coins.

Reply:
Yes, this interface is for dealing with arbitrary amounts, mostly sums (but not always).

A single decomposer should allow many decompositions to be computed, so if coins' amounts must be separated it should still be safe to share a decomposer between them (stateless), and the constructor parameters (and class) should only determine how to decompose the amounts (for example which denominations are available for use and their fees, or the cost function for optimizing decomposers).

Inputs should be decomposed separately if their link probability is high, but the goal should be to generally avoid registering multiple coins to a round when it isn't expected that the ambiguity of the corresponding input cluster will be sufficiently high for them to be decomposed as a single combined amount, because without performing this consolidation (when it's safe to do so) the wallet's UTXO pool will quickly become very fragmented.

The ambiguity of the inputs is initially based (before input registration) on privacy score of the outputs being considered for registration, with the exponential consolidation penalty applied as if it's for a standalone payment.

If the resulting score for the input cluster too low, for example when two coins are labeled and no anonymity, then their decomposition should be computed separately as fail-safe, and if the combined cost of these separate decompositions is cost-prohibitive, the inputs should not be registered into a round (perhaps until fees are lower, or the UTXO set changes). If the fail-safe decomposition(s) is/are good enough, they can be registered as inputs, and from there the situation can only improve - after connection confirmation we may learn that others' input amounts contribute enough ambiguity to the input cluster to reduce the penalty, making it safe to decompose their combined balance if it wasn't already safe to do that. Finally if the client learns about output registrations dynamically, that would provide additional opportunities to re-evaluate whether or not consolidation is possible.

For a wallet with only labeled coins which must not be linked to each other, avoiding fragmentation in all cases is not possible at first. However, after some coinjoined outputs have been created, these coins should be fine to consolidate with each other, as well as potentially with labeled coins (depending on the ambiguity these mixed coins gained). Overall approach will ensure ensure the UTXO pool of the wallet is kept down to a reasonable size without compromising privacy. It might take a while to equilibriate starting with conservative lower bounds for ambiguity metrics (the anonymity set size metric as it's currently calculated is a special case that can be generalized or a different metric could be used, I've been exploring some options), but over time we should be able to improve these estimates to reduce the overall blockspace consumption while maintaining the same level of privacy.

Later with payments from coinjoins, knapsack amount outputs and batching there's additional considerations for how to combine amounts for decomposition, but the purpose of this component is to deal with arbitrary balances, not just individual coins.